### PR TITLE
Added plot for Compare networks

### LINKED ISSUE #559: Creating semantic pointer cloud on a spa.Compare network causes error
If you try to create a Semantic Pointer Cloud on a spa.Compare network it causes an error, because `pointer.py` can't figure out the input vocabulary, because it has multiple inputs. However, this raises a few other issues. Additionally, it should be possible to get a value plot from it's output.

Firstly, neither SPA related plots should be available for that component, since it's output is a scalar and nothing useful can be shown from either the Cloud or the Plot.

Secondly, this same error will be triggered for `spa.Bind` once it's merged to master. In the `spa.Bind` case, I could fix this bug. In other words, I would recommend allowing a Semantic Pointer Cloud and even setting an override value. What do other people think?

### PR DESCRIPTION
**Pull request by Seanny123**
Before regular SPA options were given which would crash the simulation if chosen.

Resolves #559

### CODE REVIEW

File: nengo_gui/components/__init__.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from .spa_similarity import SpaSimilarity
 from .htmlview import HTMLView
 from .spike_grid import SpikeGrid
+from .comp_plot import CompPlot
```
Comment (tcstewar) [inline nengo_gui/components/__init__.py]: I assume this file was missed when you checked things in?

Comment (Seanny123) [inline nengo_gui/components/__init__.py]: ARRRRGHHH.

Fixed it.

**Comment by Seanny123** [comment]
Terry thinks sub-classing the value network is a bad idea and should be just replaced with an if-statement checking for SPA related networks. I'm thinking that maybe just checking for an "output" node might not be a bad idea and would allow for the accommodation of Ensemble Arrays and other things more easily.

**Comment by tcstewar** [comment]
> I'm thinking that maybe just checking for an "output" node might not be a bad idea and would allow for the accommodation of Ensemble Arrays and other things more easily.

Seems fair to me.  I think logic like this might work:

- if self.obj is a Node or Ensemble, use that
- if self.obj is a spa.Module and it has self.outputs['default'] set, use that
- if self.obj is a Network and it has self.output and self.output is a Node or Ensemble, use that

I guess technically we don't need the middle step in there, but it seems right to me to have the nengo_gui aware of the output default system, seeing as how it already exists.

**Comment by Seanny123** [comment]
Fixed the implementation as discussed.

**Comment by Seanny123** [comment]
@bjkomer this pull request is basic enough that I think you should be able to understand it. Basically, I enabled `netgraph.py` to tell `netgraph_item.js` that the `spa.Compare` object should have it's own special set of menus. Then I told `value.py` how to deal with a `spa.Compare` object.

File: nengo_gui/components/value.py [hunk 2]
```diff
@@ -14,7 +15,7 @@ class Value(Component):
                          min_value=-1, 
                          **Component.config_defaults)
 
-    def __init__(self, obj):
+    def __init__(self, obj, n_lines=0):
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: Why is n_lines being added as an option here?  Is this used anywhere?

File: nengo_gui/components/value.py [hunk 3]
```diff
@@ -22,8 +23,20 @@ def __init__(self, obj):
...
+        self.output = self.obj
+        if isinstance(self.obj, spa.module.Module):
+            self.output = self.obj.outputs['default'][0]
+        elif isinstance(self.obj, nengo.network.Network):
+            if self.obj.output != None:
+                self.output = self.object.output
+
         # the number of data values to send
-        self.n_lines = int(obj.size_out)
+        self.n_lines = 0
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: n_lines doesn't need to be set multiple times like this

File: nengo_gui/components/value.py [hunk 4]
```diff
@@ -22,8 +23,20 @@ def __init__(self, obj):
         # the pending data to be sent to the client
         self.data = []
 
+        # grab the output of the object
+        self.output = self.obj
+        if isinstance(self.obj, spa.module.Module):
+            self.output = self.obj.outputs['default'][0]
+        elif isinstance(self.obj, nengo.network.Network):
+            if self.obj.output != None:
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: Won't this crash if ```self.obj``` is a Network that doesn't have an attribute ```output```?  Did you mean ```if hasattr(self.obj, 'output'):```?

File: nengo_gui/components/value.py [hunk 5]
```diff
@@ -22,8 +23,20 @@ def __init__(self, obj):
         # the pending data to be sent to the client
         self.data = []
 
+        # grab the output of the object
+        self.output = self.obj
+        if isinstance(self.obj, spa.module.Module):
+            self.output = self.obj.outputs['default'][0]
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: This will crash if the module doesn't have a ```default``` output (there aren't any right now that don't but we may as well check for it now)

File: nengo_gui/components/netgraph.py [hunk 6]
```diff
@@ -469,8 +470,12 @@ def get_extra_info(self, obj):
         elif isinstance(obj, nengo.Ensemble):
             info['dimensions'] = int(obj.size_out)
             info['n_neurons'] = int(obj.n_neurons)
+        elif isinstance(obj, spa.Compare):
+            info['scalar_out'] = True
```
Comment (tcstewar) [inline nengo_gui/components/netgraph.py]: I'm okay with this special-case code for now, but we should add a note here that if we start having a lot of cases added, we should make this more generic.  For example, we could just do the same check as in value.py:

```
elif ((isinstance(obj, spa.module.Module) and obj.outputs.haskey('default')) or
      (isinstance(obj, spa.Network) and hasattr(obj, 'output'))):
             info['has_default_output'] = True
```

Hmm, and I'd also say that ```has_default_output``` may be clearer than ```scalar_out```, especially since there's nothing in this version of the code that forces it to be a scalar.....

Comment (tcstewar) [inline nengo_gui/components/netgraph.py]: Hmm, but on second thought, that more generic approach would cause ```spa.State``` to also have this line, which we don't want.  So let's stick with this special-case for now and worry about it later.   I still think the name of the flag should be changed, though.  :)

File: nengo_gui/components/value.py [hunk 7]
```diff
@@ -22,8 +23,13 @@ def __init__(self, obj):
         # the pending data to be sent to the client
         self.data = []
 
-        # the number of data values to send
-        self.n_lines = int(obj.size_out)
+        # grab the output of the object
+        self.output = self.obj
+        if isinstance(self.obj, spa.module.Module):
+            self.output = self.obj.outputs['default'][0]
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: My earlier comment on this line seems to have disappeared; here it is again:

> This will crash if the module doesn't have a default output (there aren't any right now that don't but we may as well check for it now)

File: nengo_gui/components/value.py [hunk 8]
```diff
@@ -68,7 +74,8 @@ def update_client(self, client):
     def javascript(self):
         # generate the javascript that will create the client-side object
         info = dict(uid=id(self), label=self.label,
-                    n_lines=self.n_lines, synapse=0)
+                        n_lines=self.n_lines, synapse=0)
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: spacing.  :)

File: nengo_gui/static/components/netgraph_item.js [hunk 9]
```diff
@@ -357,10 +358,14 @@ Nengo.NetGraphItem.prototype.generate_menu = function () {
         }
     }
     if (this.sp_targets.length > 0) {
-        items.push(['Semantic pointer cloud',
-                    function() {self.create_graph('Pointer', self.sp_targets[0]);}])
-        items.push(['Semantic pointer plot',
-            function() {self.create_graph('SpaSimilarity', self.sp_targets[0]);}])
+        if(!this.scalar_out) {
```
Comment (tcstewar) [inline nengo_gui/static/components/netgraph_item.js]: I'm uncomfortable with this check.  Whatever is generating sp_targets shouldn't be giving us invalid data in the first place.

Comment (Seanny123) [inline nengo_gui/static/components/netgraph_item.js]: So if sp_targets didn't return anything for scalar outputs, this check wouldn't be needed. Would that be an acceptable solution?

Comment (tcstewar) [inline nengo_gui/static/components/netgraph_item.js]: Exactly, since those aren't sp_targets as they aren't semantic pointers.   Although rather than checking for scalars, I'd do the check so that it doesn't include anything where ```vocab=None```, since if there's no vocab, it's not a semantic pointer.  (we might have future situations like ```spa.Compare``` but where there's a 2-dimensional output, for example)

**Comment by Seanny123** [comment]
Made all the changes discussed. Ready for review again.

File: nengo_gui/components/value.py [hunk 10]
```diff
@@ -76,3 +83,15 @@ def code_python_args(self, uids):
         # generate the list of strings for the .cfg file to save this Component
         # (this is the text that would be passed in to the constructor)
         return [uids[self.obj]]
+
+    @staticmethod
+    def default_output(obj):
+        """Find default output object for the input object if it exists"""
+        output = False
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: This should be ```None```, not ```False```

File: nengo_gui/components/netgraph.py [hunk 11]
```diff
@@ -469,8 +471,12 @@ def get_extra_info(self, obj):
         elif isinstance(obj, nengo.Ensemble):
             info['dimensions'] = int(obj.size_out)
             info['n_neurons'] = int(obj.n_neurons)
+        elif Value.default_output(obj):
```
Comment (tcstewar) [inline nengo_gui/components/netgraph.py]: I think this is clearer as ```elif Value.default_output(obj) is not None:```

File: nengo_gui/components/pointer.py [hunk 12]
```diff
@@ -96,6 +96,11 @@ def message(self, msg):
     @staticmethod
     def applicable_targets(obj):
         if isinstance(obj, Module):
-            return list(obj.outputs.keys())
+            targets = []
+            output_names = obj.outputs.keys()
+            for i, output in enumerate(obj.outputs.values()):
+                if output[0].size_out > 1:
+                    targets.append(output_names[i])
```
Comment (tcstewar) [inline nengo_gui/components/pointer.py]: I'm not sure about this ```size_out > 1``` check.  ```output[1] is not None``` seems much more robust to me.

Comment (tcstewar) [inline nengo_gui/components/pointer.py]: This enumerate logic also feels a bit hard to follow.  How about

```python
def applicable_targets(obj):
    targets = []
    if isinstance(obj, Module):
        for target_name, (obj, vocab) in obj.outputs.items():
            targets.append(target_name)
    return targets
```

or even

```python
def applicable_targets(obj):
    if isinstance(obj, Module):
        for target_name, (obj, vocab) in obj.outputs.items():
             yield target_name
```

File: nengo_gui/components/value.py [hunk 13]
```diff
@@ -22,8 +21,15 @@ def __init__(self, obj):
         # the pending data to be sent to the client
         self.data = []
 
+        # grab the output of the object
+        self.output = obj
+        default_out = self.default_output(self.obj)
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: I'd prefer this to be called as ```Value.default_output(self.obj)```

File: nengo_gui/components/value.py [hunk 14]
```diff
@@ -22,8 +21,15 @@ def __init__(self, obj):
...
+        if(default_out):
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: spacing and parentheses

File: nengo_gui/components/value.py [hunk 15]
```diff
@@ -22,8 +21,15 @@ def __init__(self, obj):
...
+            self.output = default_out
+
+
         # the number of data values to send
-        self.n_lines = int(obj.size_out)
+        self.n_lines = self.output.size_out
```
Comment (tcstewar) [inline nengo_gui/components/value.py]: Why get rid of the ```int``` here?  That's needed because some versions of numpy give size_outs that can't be serialized.....

File: nengo_gui/static/components/netgraph_item.js [hunk 16]
```diff
@@ -353,14 +358,14 @@ Nengo.NetGraphItem.prototype.generate_menu = function () {
             items.push(['XY-value', function() {self.create_graph('XYValue');}]);
         }
         if (this.html_node) {
-            items.push(['HTML', function() {self.create_graph('HTMLView');}])
+            items.push(['HTML', function() {self.create_graph('HTMLView');}]);
```
Comment (tcstewar) [inline nengo_gui/static/components/netgraph_item.js]: These changes and the ones below it are not part of this PR; they're part of #566

**Comment by Seanny123** [comment]
Fixed all the syntax errors. I've also surrendered myself to the merge conflicts that merging this branch will cause.

File: nengo_gui/components/tests/test_pointer.py [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+    # Only return non-scalar targets
+    model = spa.SPA()
+
+    with model:
+        model.comp = spa.Compare(4)
+        assert Pointer.applicable_targets(model.comp) == []
+
+        model.state = spa.State(4)
+        res = Pointer.applicable_targets(model.state)
+        assert res == model.state.outputs.keys()
```
Comment (tcstewar) [inline nengo_gui/components/tests/test_pointer.py]: this should be ```list(model.state.outputs.keys())``` so that it passes on Python 3 (```keys()``` in Python 3 returns an iterator, not a list)

Comment (Seanny123) [inline nengo_gui/components/tests/test_pointer.py]: Fixed.

**Comment by tcstewar** [comment]
> I've also surrendered myself to the merge conflicts that merging this branch will cause.

I suppose....  :) but please try to get into the habit of doing a ```git diff <filename.py>``` to look at your changes before committing them, and doing a ```git commit -p <filename.py>``` whenever there are changes that aren't part of whatever you're working on....  :)
